File: cli/src/main/java/ada/cli/commands/AboutCommand.java
package ada.cli.commands;

import ada.cli.output.CliOutput;
import ada.web.controllers.AboutResource;
import akka.stream.Materializer;
import akka.stream.javadsl.Source;
import lombok.*;

/**
 * @author Michael Wellner (michael.wellner@de.ibm.com)
 */
@Value
@EqualsAndHashCode
@AllArgsConstructor(staticName = "apply")
public class AboutCommand implements Runnable {

    private final CliOutput out;

    private final AboutResource resource;

    private final Materializer materializer;

    @Override
    public void run() {
        Source
            .fromPublisher(resource.getAboutStream())
            .runForeach(
                out::println,
                materializer);
    }
}

File: client/src/main/scala/ada/client/commands/about/AboutCommand.scala
package ada.client.commands.about

import ada.client.CommandContext
import ada.web.controllers.AboutResource
import akka.stream.Materializer
import akka.stream.scaladsl.Source

import scala.concurrent.ExecutionContext

case class AboutCommandImpl(
                             resource: AboutResource,
                             context: CommandContext)(
                             implicit materializer: Materializer,
                             ec: ExecutionContext) extends AboutCommand {

  override def run(): Unit = {
    val done = Source
      .fromPublisher(resource.getAboutStream)
      .runForeach(context.getOutput.println)

    done.onComplete({
      case _ => context.terminate()
    })
  }

}

object AboutCommandFactory {

  def create(resource: AboutResource,
             out: CommandContext)(
              implicit materializer: Materializer,
              ec: ExecutionContext): AboutCommand = {

    val cmd = AboutCommandImpl(resource, out)
    AboutCommandPicoDecorator.apply(cmd)
  }

}
